src/rights.jl

```diff
@@ -0,0 +1,70 @@
+const Functionality = Val
+const PATHBASE = "data/permissions/"
+
+has_access(guid::Snowflake, member::Member, task::Functionality) = has_access(guid, member.roles, task)
+
+function has_access(guid::Snowflake, roles::Vector{Snowflake}, ::Functionality{Task}) where Task
+    guildperms = PERMISSIONS[guid]
+    return !isdisjoint(guildperms[Task], roles)
+end
+
+function grant!(guid::Snowflake, role::Snowflake, ::Functionality{Task}) where Task
+    guildperms = PERMISSIONS[guid]
+    push!(guildperms[Task], role)
+    save(guid, guildperms)
+end
+
+function revoke!(guid::Snowflake, role::Snowflake, ::Functionality{Task}) where Task
+    guildperms = PERMISSIONS[guid]
+    delete!(guildperms[Task], role)
+    save(guid, guildperms)
+end
+
+"""GuildID->(Functionality->[permitted roles])"""
+const PERMISSIONS = Dict{Snowflake,Dict{Symbol,Set{Snowflake}}}()
+
+
+function load()
+    isdir(PATHBASE) || return
+    for file in readdir(PATHBASE)
+        guid = parse(Snowflake, file)
+        load(guid, checkpath=false)
+    end
+end
+
+function load(guid::Snowflake, dir=PATHBASE; checkpath=true)
+    checkpath && isdir(dirname(file)) || return
+    open(joinpath(dir, string(guid))) do io
+        fguid = read(io, Snowflake)
+        if fguid != guid
+            @info "file conflict, tried to load $guid but got $fguid"
+            return
+        end
+        while !eof(io)
+            lensym = read(io, UInt8)
+            lenroles = read(io, UInt8)
+            sym = Symbol(read(io, lensym))
+            roles = Set(reinterpret(Snowflake, read(io, lenroles)))
+            PERMISSIONS[sym] = roles
+        end
```

Comment:
I'm not a fan of building custom SerDes. We are already using StructTypes and JSON3 and that's pretty easy.  For example: 

https://github.com/Humans-of-Julia/HoJBot.jl/blob/99076ee77f84d36bef1eb25be40804212650bb68/src/command/discourse.jl#L109-L120 

If you choose to use custom data types:

https://github.com/Humans-of-Julia/HoJBot.jl/blob/99076ee77f84d36bef1eb25be40804212650bb68/src/type/discourse.jl#L59-L61